File: 202.happy-number.go
/*
 * @lc app=leetcode id=202 lang=golang
 *
 * [202] Happy Number
 */

// @lc code=start
func isHappy(n int) bool {
	res, num, record := 0, n, map[int]int{}

	for {
		for num > 0 {
			res += (num % 10) * (num % 10)
			num = num / 10
		}

		if _, ok := record[res]; !ok{
			if res == 1{
				return true
			}
			record[res] = res
			num = res
			res = 0
		}else{
			return false
		}
	}
    
}
// @lc code=end

File: 202.happy-number.cpp
/*
 * @lc app=leetcode id=202 lang=cpp
 *
 * [202] Happy Number
 */

// @lc code=start
class Solution {
 public:
  bool isHappy(int n) {
    auto get_sum = [](int n) -> int {
      int sum = 0;
      while (n) {
        int i = n % 10;
        n /= 10;

        sum += i * i;
      }
      return sum;
    };

    set<int> s;
    while (n != 1) {
      n = get_sum(n);
      if (s.count(n) != 0) return false;
      s.insert(n);
    }

    return true;
  }
};
// @lc code=end

